### PR TITLE
feat(data_hierarchy): implement changes for data type row 

### PR DESCRIPTION
**Pull request by jmurugan-fzj**
- For "Data type" row: Modified to include a combo-box (listing all doc type labels), Add, Edit and Delete buttons
- Add functionality
    - Should show the existing form with following entries:
        - Data type (LineEdit)
        - Title (LineEdit)
        - IRI (LineEdit)
        - Shortcut key combination (LineEdit)
        - Icon (LineEdit or drop-down for the list of icons?)
    - Removed "is structural type" checkbox from the form
 - Edit functionality
   - Edit form is an extension of add form, except for the data type entry which is read only. Also for "structure level type", icon and shortcut entries are read only items
  - Delete functionality
     - Similar as before, should be disabled for structural items

   - Structure level 0 & Structure level 1 are only supported, delete functionality should be disabled for both. Edit form for "structure level type" displays icon and shortcut as read only items

## Summary by Sourcery

Refactor the data hierarchy editor to separate the creation and editing of data types into distinct dialogs, enhancing the user interface and validation. Introduce a singleton class for managing QTA icons and add comprehensive tests for the new features.

New Features:
- Introduce a new dialog for editing existing data types, allowing users to modify attributes like IRI, title, shortcut, and icon.
- Add a new dialog for creating data types with fields for data type, title, IRI, shortcut, and icon.
- Implement a singleton class for managing QTA icons, ensuring only one instance is used throughout the application.

Enhancements:
- Refactor the data hierarchy editor to separate the creation and editing of data types into distinct dialogs.
- Improve the user interface by adding tooltips and disabling certain fields when editing existing data types.
- Enhance the validation of data type information using a dedicated validator class.

Tests:
- Add comprehensive unit tests for the new create and edit type dialogs, covering various scenarios and edge cases.
- Introduce tests for the QTAIconsSingleton class to ensure correct initialization and icon management.
- Implement tests for the DataTypeInfo and its validator to verify correct behavior and error handling.

_**Additional Info: Pyside6 version has been fixed to 6.7.3 since there is some incompatibility between the dependent libraries with the latest PySide6 6.8.0, The issue is already backlogged: [PASTA APP crashes with segmentation fault](https://github.com/PASTA-ELN/pasta-eln/issues/352)**_

### CODE REVIEW

File: pasta_eln/GUI/data_hierarchy/data_hierarchy_editor_dialog.py [hunk 1]
```diff
@@ -124,7 +121,10 @@ def __init__(self, database: Database) -> None:
     self.typeAttachmentsTableView.horizontalHeader().setSectionResizeMode(1, QtWidgets.QHeaderView.ResizeMode.Stretch)
 
     # Create the dialog for new type creation
-    self.create_type_dialog = CreateTypeDialog(self.create_type_accepted_callback, self.create_type_rejected_callback)
+    self.create_type_dialog: TypeDialog = CreateTypeDialog(self.type_create_accepted_callback,
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply pasta_eln/GUI/data_hierarchy/data_hierarchy_editor_dialog.py:124]: Already done!

File: tests/unit_tests/test_data_hierarchy_edit_type_dialog.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        )
+        mock_generate_data_hierarchy_type.assert_called_once_with(edit_type_dialog.type_info)
+        edit_type_dialog.selected_data_hierarchy_type.update.assert_called_once_with({"key": "updated_value"})
+        edit_type_dialog.instance.close.assert_called_once()
+        edit_type_dialog.accepted_callback_parent.assert_called_once()
+      else:
+        edit_type_dialog.logger.info.assert_not_called()
+        mock_generate_data_hierarchy_type.assert_not_called()
+        if selected_data_hierarchy_type:
+          edit_type_dialog.selected_data_hierarchy_type.update.assert_not_called()
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply tests/unit_tests/test_data_hierarchy_edit_type_dialog.py:289]: NA

File: tests/unit_tests/test_data_hierarchy_edit_type_dialog.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        edit_type_dialog.instance.close.assert_not_called()
+        edit_type_dialog.accepted_callback_parent.assert_not_called()
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply tests/unit_tests/test_data_hierarchy_edit_type_dialog.py:291]: NA

File: tests/unit_tests/test_data_hierarchy_edit_type_dialog.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+                             expected_message, test_id):
+    # Arrange
+    edit_type_dialog.type_info = MagicMock(datatype="test_datatype", title="test_title")
+    edit_type_dialog.validate_type_info = MagicMock()
+    edit_type_dialog.validate_type_info.return_value = validate_type_info
+    if selected_data_hierarchy_type:
+      edit_type_dialog.selected_data_hierarchy_type = MagicMock()
+      edit_type_dialog.selected_data_hierarchy_type.__getitem__.side_effect = selected_data_hierarchy_type.__getitem__
+    else:
+      edit_type_dialog.selected_data_hierarchy_type = None
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply tests/unit_tests/test_data_hierarchy_edit_type_dialog.py:266]: NA

File: tests/unit_tests/test_data_hierarchy_edit_type_dialog.py [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+      edit_type_dialog.typeDisplayedTitleLineEdit.setText.assert_called_once_with(
+        edit_type_dialog.selected_data_hierarchy_type["title"] or "")
+      edit_type_dialog.shortcutLineEdit.setText.assert_called_once_with(
+        edit_type_dialog.selected_data_hierarchy_type["shortcut"] or "")
+      edit_type_dialog.iconFontCollectionComboBox.setCurrentText.assert_called_once_with(
+        edit_type_dialog.selected_data_hierarchy_type["icon"].split(".")[0] if
+        edit_type_dialog.selected_data_hierarchy_type[
+          "icon"] else "")
+      edit_type_dialog.iconComboBox.setCurrentText.assert_called_once_with(
+        edit_type_dialog.selected_data_hierarchy_type["icon"] or "No value")
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply tests/unit_tests/test_data_hierarchy_edit_type_dialog.py:236]: NA

File: tests/unit_tests/test_data_hierarchy_data_type_info_validator.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    # Act
+    if expected_exception:
+      # Assert
+      with pytest.raises(expected_exception):
+        DataTypeInfoValidator.validate(data_type_info)
+    else:
+      # Act
+      DataTypeInfoValidator.validate(data_type_info)
+      # Assert
+      # No exception should be raised
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply tests/unit_tests/test_data_hierarchy_data_type_info_validator.py:58]: NA

File: pasta_eln/dataverse/data_type_info.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+    Raises:
+        IncorrectParameterError: If the provided shortcut is not a string or None.
+    """
+    if isinstance(shortcut, str | None):
+      self._shortcut = shortcut
+    else:
+      raise IncorrectParameterError(f"Expected string type for shortcut but got {type(shortcut)}")
+
+  def __iter__(self) -> Generator[tuple[str, Any], None, None]:
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply pasta_eln/dataverse/data_type_info.py:219]: vars(self) does not relaibly returns the the public properties of an object, neither does others! dir() allows the return but also provides the methods too!

File: pasta_eln/GUI/data_hierarchy/data_hierarchy_editor_dialog.py [hunk 8]
```diff
@@ -305,38 +281,63 @@
...
-    self.typeDisplayedTitleLineEdit.textChanged[str].connect(self.update_type_displayed_title)
-    self.typeIriLineEdit.textChanged[str].connect(self.update_type_iri)
-
     self.metadataGroupComboBox.clear()
     self.addMetadataGroupLineEdit.clear()
     self.typeMetadataTableView.model().update([])
     self.typeAttachmentsTableView.model().update([])
 
-  def create_type_accepted_callback(self) -> None:
+  def type_create_accepted_callback(self) -> None:
```
[the remark this answers is not in the document]
Comment (jmurugan-fzj) [reply pasta_eln/GUI/data_hierarchy/data_hierarchy_editor_dialog.py:289]: Not needed!

**Comment by jmurugan-fzj** [comment]
@SteffenBrinckmann Could you please do the review, had fixed all the failing problems and addressed valid comments from Sourcery too!

**Review by SteffenBrinckmann — APPROVED** [review]
@jmurugan-fzj Pyside6.8.0.1 was released today, just 5 days after 6.8.0 and they usually do not release 4-digit releases. It makes me think, did the issue you encounter, occur to them too and they have fixed it. Can you try to relax the requirements (without =6.7.3) and see if it works now?
